Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Hole_filling/Triangulate_hole_polyline.h

```diff
@@ -1194,7 +1203,176 @@ class Triangulate_hole_polyline {
+// /!\ points.first == points.last
+
+
+template <
+  typename PointRange, //need size()
+  typename Tracer,
+  typename Validity_checker,
+  typename Traits
+>
+bool
+triangulate_hole_polyline_with_cdt(const PointRange& points,
+                          Tracer& tracer,
+                          const Validity_checker& is_valid,
+                          const Traits& traits)
+{
+  typedef typename Traits::FT FT;
+  typedef typename Traits::Point_3 Point_3;
+  typedef typename Traits::Vector_3 Vector_3;
+  typedef typename Traits::Collinear_3 Collinear_3;
+  typedef typename Traits::Compute_squared_length_3 Squared_length_3;
+
+  // Compute an average normal of the hole.
+  const Squared_length_3 squared_length_3 =
+    traits.compute_squared_length_3_object();
+  const Collinear_3 collinear_3 =
+    traits.collinear_3_object();
+
+  std::vector<Point_3> P(boost::begin(points), boost::end(points));
+  if(P.front() != P.back()){
+    P.push_back(P.front());
+  }
+
+  FT x = FT(0), y = FT(0), z = FT(0);
+  std::size_t num_normals = 0;
+  const Point_3& ref_point = P[0];
+  std::size_t size = P.size()-1;
+  for (std::size_t i = 1; i < size; ++i) {
+    const std::size_t ip = i + 1;
+
+    const Point_3& p1 = ref_point; // 3 points, which form a triangle
+    const Point_3& p2 = P[i];
+    const Point_3& p3 = P[ip];
+
+    // Skip in case we have collinear points.
+    if (collinear_3(p1, p2, p3))
+      continue;
+
+    // Computing the normal of a triangle.
+    Vector_3 tri_normal = CGAL::normal(p1, p2, p3);
+    const FT tri_normal_length = static_cast<FT>(
+      CGAL::sqrt(CGAL::to_double(squared_length_3(tri_normal))));
+    if(tri_normal_length <= FT(0))
+      return false;
+
+    tri_normal /= tri_normal_length;
+
+    x += tri_normal.x(); y += tri_normal.y(); z += tri_normal.z();
+    ++num_normals;
+  }
+
+  if(num_normals < 1)
+    return false;
+
+  x /= static_cast<FT>(num_normals);
+  y /= static_cast<FT>(num_normals);
+  z /= static_cast<FT>(num_normals);
+
+  // Setting the final normal.
+  const Vector_3 normal = Vector_3(x, y, z);
+  const FT normal_length = static_cast<FT>(
+    CGAL::sqrt(CGAL::to_double(squared_length_3(normal))));
+  const Vector_3 avg_normal = normal / normal_length;
+
+  // Checking the hole simplicity.
+  typedef Triangulation_2_projection_traits_3<Traits> P_traits;
+  const P_traits p_traits(avg_normal);
+  if(!is_simple_2(P.begin(), P.end()-1, p_traits))
+     return false;
+
+
+  Lookup_table_map<int> lambda(static_cast<int>(size),-1);
+  //Create and fill the cdt_2
+  typedef CGAL::Triangulation_vertex_base_with_info_2<std::size_t, P_traits> Vb;
+  typedef CGAL::Triangulation_face_base_with_info_2<bool, P_traits>          Fb1;
+  typedef CGAL::Constrained_triangulation_face_base_2<P_traits, Fb1>         Fb;
+  typedef CGAL::Triangulation_data_structure_2<Vb,Fb>                        TDS;
+  typedef CGAL::Exact_intersections_tag                                      Itag;
```

Comment:
if we keep the `is_simple()` test above I think we should not allow intersection of constraints (because they should simply never happen). I would update that tag then.